workflow-manager/task/task.go

```diff
@@ -0,0 +1,253 @@
+}
+
+type IntakeBatch struct {
+	// AggregationID is the identifier for the aggregaton
+	AggregationID string `json:"aggregation-id"`
+	// BatchID is the identifier of the batch. Typically a UUID.
+	BatchID string `json:"batch-id"`
+	// Date is the timestamp on the batch
+	Date Timestamp `json:"date"`
+}
+
+func (i IntakeBatch) Marker() string {
+	return fmt.Sprintf("intake-%s-%s-%s", i.AggregationID, i.Date.MarkerString(), i.BatchID)
+}
+
+// Enqueuer allows enqueuing tasks.
+type Enqueuer interface {
+	// Enqueue enqueues a task to be executed later. The provided completion
+	// function will be invoked once the task is either successfully enqueued or
+	// some unretryable error has occurred. A call to Stop() will not return
+	// until completion functions passed to any and all calls to Enqueue() have
+	// returned.
+	Enqueue(ctx context.Context, task Task, completion func(error))
+	// Stop blocks until all tasks passed to Enqueue() have been enqueued in the
+	// underlying system, and all completion functions pased to Enqueue() have
+	// returned, and so it is safe to exit the program without losing any tasks.
+	Stop()
+}
+
+// CreatePubSubTopic creates a PubSub topic with the provided ID, as well as a
+// subscription with the same ID that can later be used by a facilitator.
+// Returns error on failure.
+func CreatePubSubTopic(project string, topicID string) error {
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+	client, err := pubsub.NewClient(ctx, project)
+	if err != nil {
+		return fmt.Errorf("pubsub.newClient: %w", err)
+	}
+
+	topic, err := client.CreateTopic(ctx, topicID)
+	if err != nil {
+		return fmt.Errorf("pubsub.CreateTopic: %w", err)
+	}
+
+	tenMinutes, _ := time.ParseDuration("10m")
+
+	subscriptionConfig := pubsub.SubscriptionConfig{
+		Topic:            topic,
+		AckDeadline:      tenMinutes,
+		ExpirationPolicy: time.Duration(0), // never expire
+	}
+	if _, err := client.CreateSubscription(ctx, topicID, subscriptionConfig); err != nil {
+		return fmt.Errorf("pubsub.CreateSubscription: %w", err)
+	}
+
+	return nil
+}
+
+// GCPPubSubEnqueuer implements Enqueuer using GCP PubSub
+type GCPPubSubEnqueuer struct {
+	topic     *pubsub.Topic
+	waitGroup sync.WaitGroup
+	dryRun    bool
+}
+
+// NewGCPPubSubEnqueuer creates a task enqueuer for a given project and topic
+// in GCP PubSub. If dryRun is true, no tasks will actually be enqueued. Clients
+// should re-use a single instance as much as possible to enable batching of
+// publish requests.
+func NewGCPPubSubEnqueuer(ctx context.Context, project string, topicID string, dryRun bool) (*GCPPubSubEnqueuer, error) {
+	client, err := pubsub.NewClient(ctx, project)
+	if err != nil {
+		return nil, fmt.Errorf("pubsub.NewClient: %w", err)
+	}
+
+	return &GCPPubSubEnqueuer{
+		topic:  client.Topic(topicID),
+		dryRun: dryRun,
+	}, nil
+}
+
+func (e *GCPPubSubEnqueuer) Enqueue(ctx context.Context, task Task, completion func(error)) {
```

Comment:
You use an external ctx if you want to group multiple operations in one ctx that you can cancel if any of the participants errors and thus release resources in all participating systems. If there is no grouping of operations is required you can create the context here in the function and don't need to pass it as a fn param.

```
ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
defer cancel()
```